### PR TITLE
ENH: Quantile Forest Support

### LINKED ISSUE #383: Can't Load Quantile Forest
I'm currently using a Quantile Forest using the zillow library (https://github.com/zillow/quantile-forest/tree/efc1ebd08f1431a33532a418896cc14f22e22be3). I was able to dump the model just fine but am seeing an issue when attempting to load the model.  I've set the `trusted` parameter to true but I get the following error: 

```
ObjectNode._construct(self)
    408 cls = gettype(self.module_name, self.class_name)
    410 # Instead of simply constructing the instance, we use __new__, which
    411 # bypasses the __init__, and then we set the attributes. This solves the
    412 # issue of required init arguments. Note that the instance created here
    413 # might not be valid until all its attributes have been set below.
--> 414 instance = cls.__new__(cls)  # type: ignore
    416 if not self.children["attrs"]:
    417     # nothing more to do
    418     return instance

File quantile_forest/_quantile_forest_fast.pyx:568, in quantile_forest._quantile_forest_fast.QuantileForest.__cinit__()

TypeError: __cinit__() takes at least 2 positional arguments (0 given)
```

This seems to occur when attempting to instantiate the model type when validating/creating appropriate attributes: https://github.com/skops-dev/skops/blob/main/skops/io/_general.py#L414.  The model object type requires 2 positional arguments (https://github.com/zillow/quantile-forest/blob/efc1ebd08f1431a33532a418896cc14f22e22be3/quantile_forest/_quantile_forest_fast.pyx#L568) that aren't being passed in, and there seems to not be a straight forward way to pass them in.

I was wondering if there might be some workaround that I'm not seeing. At the very least, it seems unusual to allow for a model type to be dumped without validating that it can be opened.  Meaning, should the dump functionality check for this initialization before attempting to save in order to ensure the object that is being persisted can indeed be loaded.

### PR DESCRIPTION
**Pull request by reidjohnson**
Fixes #383.

The [quantile-forest](https://github.com/zillow/quantile-forest) package extends sklearn random forest regressors to implement quantile regression forests. Without these changes, the Cython-based model object is unable to be loaded by skops due to its use of the `__reduce__` method.

I understand adding support for another package may be outside the current scope of this project. As the primary maintainer of the quantile-forest package, I'd also be interested in better understanding if there's a recommended alternative to the use of `__reduce__` that would enable the model to be compatible with this project's existing secure persistence feature.

### CODE REVIEW

File: skops/_min_dependencies.py
Comment (BenjaminBossan) [inline skops/_min_dependencies.py:18]:
```suggestion
    "quantile-forest": ("1.0.0", "tests", None),
```

**Review by BenjaminBossan — COMMENTED** [review]
Thanks so much for providing this fix. I haven't done a proper review yet. Would you mind adding a test for quantile forest estimators? It could be similar to the tests [here](https://github.com/skops-dev/skops/blob/main/skops/io/tests/test_external.py). If you need help with that, let us know.

File: skops/io/_visualize.py
Comment (reidjohnson) [inline skops/io/_visualize.py:235]: Suggested resolution for #385.

**Comment by reidjohnson** [comment]
@BenjaminBossan: Sure, I took a pass at adding a test for the quantile forest estimators. This also uncovered a bug in the visualization of sklearn DecisionTreeRegressors reported as #385, with a suggested resolution included in these changes.

File: skops/io/_quantile_forest.py
Comment (adrinjalali) [inline skops/io/_quantile_forest.py:38]: I'd be happier if we don't trust them by default.

File: skops/io/_visualize.py
Comment (adrinjalali) [reply skops/io/_visualize.py:235]: I would include that in a separate PR.

File: skops/io/tests/test_external.py
Comment (adrinjalali) [inline skops/io/tests/test_external.py:415]: move to `@pytest.mark.parametrize`?

**Review by adrinjalali — COMMENTED** [review]
Thanks for the nice PR. Other than the small nits, I think this looks good.

Might be time for us to start working on a better way to handle C extension types.

File: skops/io/_quantile_forest.py
Comment (reidjohnson) [reply skops/io/_quantile_forest.py]: Updated.

File: skops/io/tests/test_external.py
Comment (reidjohnson) [reply skops/io/tests/test_external.py]: Updated.

**Comment by adrinjalali** [comment]
@reidjohnson tests failing

**Comment by reidjohnson** [comment]
Failing due to #385.

File: skops/io/_persist.py
Comment (adrinjalali) [reply skops/io/_persist.py:18]: Right, good catch. What I think we should do is to delay required import to when the user dumps, at which point if they have an instance from this library, they should have the library installed. And then when loading, the same, we try to load the module / type, and if it doesn't exist we raise.

Comment (BenjaminBossan) [reply skops/io/_persist.py:18]: Is the issue not more during loading though? We need to import the class because `get_state` dispatches based on it. Therefore, we cannot lazily load `QuantileForest` when the object is constructed. Otherwise, I would agree.

Comment (adrinjalali) [reply skops/io/_persist.py:18]: Loading depends on dispatch based on the dictionaries we're setting based on the `__loader__`, not type, so that dispatch is fine and can delay any imports to when they're needed, I think?

Comment (BenjaminBossan) [reply skops/io/_persist.py:18]: Ah yes, sorry, you're right, it's needed during dumping, not loading. So is your suggestion that this code:

https://github.com/skops-dev/skops/blob/923dafb64433f599b2f62f8b746938979e1aea03/skops/io/_persist.py#L20-L26

should not be called on the module root, but instead the first time that `dump` / `dumps` is being called?

Comment (adrinjalali) [reply skops/io/_persist.py:18]: what we could do, is inside `_quantile_forest.py` only do lazy imports, and only fill the dictionary if the library is installed, that should fix the issue I think.

Comment (BenjaminBossan) [reply skops/io/_persist.py:18]: That should work too.

Comment (reidjohnson) [reply skops/io/_persist.py:18]: Thanks for the conversation; updated with lazy imports inside `_quantile_forest.py`.

File: skops/io/_quantile_forest.py
Comment (adrinjalali) [inline skops/io/_quantile_forest.py:49]: What do you think of this instead? (slightly different)

```python
from __future__ import annotations

from typing import Any, Sequence

from ._protocol import PROTOCOL
from ._sklearn import ReduceNode, reduce_get_state
from ._utils import LoadContext, SaveContext

try:
    from quantile_forest._quantile_forest_fast import QuantileForest
except ImportError:
    QuantileForest = None


def quantile_forest_get_state(
    obj: Any,
    save_context: SaveContext,
) -> dict[str, Any]:
    state = reduce_get_state(obj, save_context)
    state["__loader__"] = "QuantileForestNode"
    return state


class QuantileForestNode(ReduceNode):
    def __init__(
        self,
        state: dict[str, Any],
        load_context: LoadContext,
        trusted: bool | Sequence[str] = False,
    ) -> None:
        if QuantileForest is None:
            raise ImportError(
                "`quantile_forest` is missing and needs to be installed in order to"
                " load this model."
            )

        super().__init__(
            state,
            load_context,
            constructor=QuantileForest,
            trusted=trusted,
        )
        self.trusted = self._get_trusted(trusted, [])


# tuples of type and function that gets the state of that type
if QuantileForest is not None:
    GET_STATE_DISPATCH_FUNCTIONS = [(QuantileForest, quantile_forest_get_state)]

# tuples of type and function that creates the instance of that type
NODE_TYPE_MAPPING = {("QuantileForestNode", PROTOCOL): QuantileForestNode}
```

**Review by adrinjalali — APPROVED** [review]

**Review by BenjaminBossan — APPROVED** [review]
LGTM, thanks a lot for adding the support. I have an extremely minor comment, but this is good to be merged from my point of view.

File: skops/io/tests/test_external.py
Comment (reidjohnson) [reply skops/io/tests/test_external.py:387]: These are the user-facing classes which are explicitly tested, `QuantileForest` is the underlying Cython object shared by the user-facing classes and isn't explicitly tested here (but is serialized by the loading/dumping).

**Comment by adrinjalali** [comment]
I just noticed this is missing a changelog entry.

**Comment by reidjohnson** [comment]
Updated, thanks for flagging.

**Comment by brentonmallen1** [comment]
awesome work everyone, y'all are amazing!  I really appreciate it!

**Comment by brentonmallen1** [comment]
I finally got back to this effort and it seems like, even with the changes, I'm still seeing the following error when attempting to load a quantile forest model: 

```line 568, in quantile_forest._quantile_forest_fast.QuantileForest.__cinit__
TypeError: __cinit__() takes at least 2 positional arguments (0 given)```

Did this come up during testing? Am I missing something?

**Comment by brentonmallen1** [comment]
more specifically: `File "quantile_forest/_quantile_forest_fast.pyx", line 568, in quantile_forest._quantile_forest_fast.QuantileForest.__cinit__`

**Comment by reidjohnson** [comment]
@brentonmallen1 Just checking, but are you using the development build? I don't believe there's yet been a public release since these changes were merged.

**Comment by brentonmallen1** [comment]
I was using the main release for 0.8.0 because the changelog had mentioned QuantileForest support, but I could have misread. I can try using the dev release and give it another go and report back. thank you

**Comment by brentonmallen1** [comment]
well, I don't see the quantile changes in my site-packages so it does seem like I haven't installed what I was expecting. I also don't see a dev branch/build, how would I go about installing that directly?

**Comment by reidjohnson** [comment]
@brentonmallen1 I'll defer to the maintainers, but a viable route is to follow the dev setup described in the contributing doc.

**Comment by adrinjalali** [comment]
We just released 0.9.0, which includes this change.

**Comment by brentonmallen1** [comment]
y'all are wonderful! thank you!

**Pull request merged**
